### PR TITLE
[WIP] #356 Post Consents

### PR DESCRIPTION
**Pull request by akapur99**
PR for ticket 356 on Zenhub: https://app.zenhub.com/workspaces/pisp-5e8457b05580fb04a7fd4878/issues/mojaloop/mojaloop/356
Deals with (and have completed): 
1. Request handler for POST /consents
2. Validation of inbound request
3. Create Consent object and store in database
4. Create scope(s) object and store in database
5. Unit tests for above

BDD and Integration testing done in another ticket (No. 354)

### CODE REVIEW

File: src/server/handlers/consents.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    ),
+    credential: Joi.string().validate(null)
+  })
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  const { error, value } = schema.validate(request.payload)
+  if (error) {
+    Logger.push(error).error('Error: Invalid Request ')
+    return h.response().code(400)
+  }
```
Comment (jgeewax) [inline src/server/handlers/consents.ts:66]: This is probably handled already by the OAS doc.

File: src/server/handlers/consents.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+  // Asynchronously deals with creation and storing of consents and scope
+  setImmediate(async (): Promise<void> => {
+    try {
+      await createAndStoreConsent(request)
+    } catch (error) {
+      Logger.push(error).error('Error: Unable to create/store consent')
+      throw error
```
Comment (jgeewax) [inline src/server/handlers/consents.ts:74]:
```suggestion
      // TODO: Decide what to do with this error.
```

File: src/server/handlers/consents.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  const { error, value } = schema.validate(request.payload)
+  if (error) {
+    Logger.push(error).error('Error: Invalid Request ')
+    return h.response().code(400)
+  }
+
+  // Asynchronously deals with creation and storing of consents and scope
+  setImmediate(async (): Promise<void> => {
```
Comment (jgeewax) [inline src/server/handlers/consents.ts:69]: https://jestjs.io/docs/en/timer-mocks

Comment (eoln) [reply src/server/handlers/consents.ts:53]: @jgeewax timer mocks doesn't mock `setImmediate`. 
@akapur99 better is in test mock the `../domain/consents` to deliver mocked `createAndStoreConsent` function. At least two unit tests you should deliver one for happy flow and second for throwing the exception. Use setTimeout to allow `setImmediate` code to be executed

File: test/unit/server/handlers/consents.test.ts [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    mockStoreConsent.mockResolvedValue()
+  })
+
+  it('Should return 202 success code', async (): Promise<void> => {
+    const response = await post(
+      request as Request,
+      h as ResponseToolkit
+    )
+    expect(response).toBe(h.response().code(202))
+    expect(mockStoreConsent).toHaveBeenCalled()
```
Comment (eoln) [inline test/unit/server/handlers/consents.test.ts:132]: you should use `toHaveBeenCalledWith`

File: test/unit/server/handlers/consents.test.ts [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+  })
+
+  it('Should return 400 code due to invalid request', async (): Promise<void> => {
+    const response = await post(
+      requestInvalid as Request,
+      h as ResponseToolkit
+    )
+    expect(response).toBe(h.response().code(400))
+
+    expect(mockStoreConsent).not.toHaveBeenCalled()
```
Comment (eoln) [inline test/unit/server/handlers/consents.test.ts:142]: you should use `toHaveBeenCalledWith`

File: test/unit/server/domain/consents.test.ts [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+  beforeAll((): void => {
+    mockRegisterConsent.mockResolvedValue(null)
+    mockRegisterScopes.mockResolvedValue(null)
+  })
+  it('Should return nothing and no errors thrown', async (): Promise<void> => {
+    expect(async (): Promise<void> => {
+      await createAndStoreConsent(request)
+    }).not.toThrowError()
+
+    expect(mockRegisterConsent).toHaveBeenCalled()
```
Comment (eoln) [inline test/unit/server/domain/consents.test.ts:73]: you should use `toHaveBeenCalledWith`

**Comment by eoln** [comment]
@akapur99 Is using <> brackets in the commit message to wrap the type compliant with https://www.conventionalcommits.org/en/v1.0.0/#specification?

File: src/domain/consents.ts [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+  initiatorId: string;
+  participantId: string;
+  scopes: ExternalScope[];
+  credential: null;
+}
+
+/**
+   * Validates whether request is valid
+   * by comparing if source header matches participant ID
+   * @param request: request received from switch
```
Comment (ramanmangla) [inline src/domain/consents.ts:57]: You might want to resolve the inconsistency in function docs.

File: test/unit/lib/scopes.test.ts [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+  accountId: 'as2342',
+  actions: ['account.getAccess', 'account.transferMoney']
+},
+{
+  accountId: 'as22',
+  actions: ['account.getAccess']
+}
+]
+
+// TODO: Add more tests
```
Comment (lewisdaly) [inline test/unit/lib/scopes.test.ts:63]: Please don't leave todos in here ;)

File: test/unit/lib/scopes.test.ts [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+describe('Scope Convert ExternalScope to Scope', (): void => {
+  it('Should return Scope array when input ExternalScope array', (): void => {
+    expect(ScopeFunctions.convertExternalToScope(externalScope, consentId)).toStrictEqual(scopes)
```
Comment (lewisdaly) [inline test/unit/lib/scopes.test.ts:66]: I generally like to split tests into more lines to make them more readable, but it's just stylistic.

File: test/unit/server/handlers/consents.test.ts [hunk 10]
```diff
@@ -163,7 +172,7 @@ describe('server/handlers/consents', (): void => {
       expect(response).toBe(202)
       jest.runAllImmediates()
 
-      expect(setImmediate).rejects.toThrowError()
+      // expect(setImmediate).rejects.toThrowError()
```
Comment (eoln) [inline test/unit/server/handlers/consents.test.ts]: why is this code commented?

File: test/unit/lib/scopes.test.ts (hunk 8 above)
Comment (akapur99) [reply test/unit/lib/scopes.test.ts:63]: oops

File: test/unit/server/handlers/consents.test.ts (hunk 10 above)
Comment (akapur99) [reply test/unit/server/handlers/consents.test.ts]: ah sorry that comment should be removed - setImmediate itself will not be throwing an error as the try catch exists within the callback function within it

File: src/lib/scopes.ts [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+
+/**
+ * Takes input of array of ExternalScope objects
+ * Reformats and returns array of Scope objects
+ * @param externalScopes Array of ExternalScope objects received
+ * @param consentId Id of Consent to which scopes belong
+ */
+export function convertExternalToScope (
+  externalScopes: ExternalScope[], consentId: string): Scope[] {
+  const scopes: Scope[] = []
```
Comment (eoln) [inline src/lib/scopes.ts:49]:
```suggestion
  let scopes: Scope[] = []
```

File: src/domain/consents.ts [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+  const consent: Consent = {
+    id: payload.id,
+    initiatorId: payload.initiatorId,
+    participantId: payload.participantId
+  }
+
+  const scopes: Scope[] = convertExternalToScope(payload.scopes, consent.id)
+
+  try {
+    await consentDB.insert(consent)
```
[the remark this answers is not in the document]
Comment (akapur99) [reply src/domain/consents.ts:81]: not sure - @jgeewax  @ramanmangla  do you think it should be in the sql transaction?
Per our discussion shouldn't this be in domain/handler logic

File: src/lib/scopes.ts [hunk 13]
```diff
@@ -46,24 +47,15 @@ export interface ExternalScope {
...
-     (element: ExternalScope): Scopes[] => element.actions.map((action: string): Scope => ({
-            consentId,
-            accountId: element.accountId,
-            action
-     })
-  ))
-    const accountId = element.accountId
-    element.actions.forEach((action: string): void => {
-      const scope = {
+  const scopes: Scope[] = ([] as Scope[]).concat(...externalScopes.map(
```
Comment (eoln) [inline src/lib/scopes.ts:50]: I think we can still make it simpler, we can replace `concat` and `spread opeartor` with `flat`
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flat

```suggestion
  const scopes: Scope[] = externalScopes.map(
```

File: src/lib/scopes.ts [hunk 14]
```diff
@@ -46,24 +47,15 @@ export interface ExternalScope {
...
-        accountId,
+        accountId: element.accountId,
         action
-      }
-      scopes.push(scope)
-    })
-  })
+      })
+      )
+  ))
```
Comment (eoln) [inline src/lib/scopes.ts:58]:
```suggestion
  ).flat()
```
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flat

**Pull request merged**